### PR TITLE
Hole LEM PR 

### PR DESCRIPTION
**Pull request by sjjs7**
Final PR with all current changes. 
Fixed up indenting on a number of files as well (will be more mindful of that- sorry!).

### CODE REVIEW

File: Constructions/QuotationTactics.ml [hunk 1]
```diff
@@ -1,29 +1,188 @@
...
 
-let rec STRING_FETCH tm = match tm with
-	| Comb(Comb(Const("=",_),a),b) when type_of a = mk_type("list",[mk_type ("char",[])]) -> [STRING_EQ_CONV(tm)]
-	| Comb(a,b) -> (STRING_FETCH a) @ (STRING_FETCH b)
-	| Abs(a,b) -> (STRING_FETCH a) @ (STRING_FETCH b)
-	| Quote(e) -> (STRING_FETCH e)
-	| Hole(e,ty) -> (STRING_FETCH e)
-	| Eval(e,ty) -> (STRING_FETCH e)
-	| _ -> [];; 
+let rec STRING_FETCH tm =
```
Comment (JacquesCarette) [inline Constructions/QuotationTactics.ml:7]: I'm concerned about the need for this. Why are 'strings' being manipulated like this at all? [There might be good reasons, but just as often, it's a sign of hacking]

**Comment by JacquesCarette** [comment]
Otherwise, everything here seemed fine to me. But there are definitely some 'deep' QE things going on here that @wmfarmer needs to check in detail.

File: Constructions/QuotationTactics.ml (hunk 1 above)
Comment (sjjs7) [reply Constructions/QuotationTactics.ml:7]: This function/tactic is meant to evaluate the equating of strings to T or F.
For example, in epsilon.ml, isVar is defined by checking that the constructor is "QuoVar" which requires equating strings.
It's just turning Harrison's STRING_EQ_CONV into a tactic which checks the whole term over for terms of type "s" = "t" and simplifies them, I didn't think anything maleficent was going on.

Was my change the problem?

**Comment by sjjs7** [comment]
@wmfarmer we've gone through most of these changes, but take a final look.

File: Constructions/QuotationTactics.ml (hunk 1 above)
Comment (JacquesCarette) [reply Constructions/QuotationTactics.ml:7]: Internally, you have to have some code manipulate low-level representations. But only internal routines should ever call such low-level manipulations. This shouldn't leak out to the user. [ And if STRING_EQ_CONV is user-visible, I would consider that to be a bug on Harrison's part]

If it's just about T and F, maybe it's better to hard-code that, than to let some general name-revealing routines leak out too far.

Comment (sjjs7) [reply Constructions/QuotationTactics.ml:7]: Ok I will hard-code it in.

**Review by wmfarmer — APPROVED** [review]
I only noticed a couple things that should be changed:

1.  "is_quasi_construction" should be called  "is_construction" since we are allowing constructions to contain arbitrary subterms of type epsilon.
2. The eval case in "CONSTRUCTION_TO_QUOTE_CONV" should be
       | Eval(e,ty) -> Eval(ctq e, ty)

File: Constructions/QuotationTactics.ml (hunk 1 above)
Comment (sjjs7) [reply Constructions/QuotationTactics.ml:7]: This is going to take longer than I thought- lots of the functions are defined in terms of comparing strings. 
Most of the changes will result in no longer being able to prove formulas false- which I guess is fine, I'll just remove the tests that evaluate to F.
I am unsure how to deal with the isFreeIn theorem though, as it requires checking that the variable is not the binding variable in an abstraction, which I can no longer do if I don't compare strings. 

Am I allowed to change Harrison's original tactics to automatically equate strings (for example REWRITE_TAC)? 
I feel like that's cheating though...
I'm not sure how to get around that issue.

Comment (JacquesCarette) [reply Constructions/QuotationTactics.ml:7]: Defining some functions in terms of comparing strings is fine! What is not good is letting the implementations details leak.  So as long as `STRING_FETCH` is not user-accessible, that's fine.

Comment (sjjs7) [reply Constructions/QuotationTactics.ml:7]: But if I made STRING_FETCH an internal function in another function (like REWRITE_TAC for ex.) isn't it still kind of "user-accessible"? Or is that not the right path.

I'm confused about how to do this.
(also, is the plan to never use STRING_EQ_CONV in a proof?)

Comment (wmfarmer) [reply Constructions/QuotationTactics.ml:7]: Values of type "string" are components of both types and terms.  So the proof developer needs to be able to access them. In this case, "string" is an external representation.

Comment (JacquesCarette) [reply Constructions/QuotationTactics.ml:7]: I think that is too simplistic a view. It's like saying that organic molecules (eventually) contain quarks, and so these need to be accessible to do anything.

The proof developer should never need to see the implementation type "string" -- if they do, I would consider that a very severe bug. They may need to know that certain things have 'names' and that 'names' have an API. But that API should be very poor (a tiny fraction of the things you can actually do with strings) and abstract, in that string operations cannot be applied to names.

The whole point of the ML-HOL approach to theorem proving is that 'thm' is an abstract type.  While 'type' and 'term' are not quite as abstract, still, they do not expose their guts for all to see and play with.

**Comment by sjjs7** [comment]
This is good to commit.

File: fusion.ml [hunk 2]
```diff
@@ -1141,19 +1141,19 @@ let rec type_subst i ty =
...
       | Comb(Comb(Const("QuoVar",_),_),_) -> true
       | Comb(Comb(Const("QuoConst", _), _), _) -> true
-      | Comb(Const("Quo", ty), e) -> is_quasi_construction e
-      | Comb(Comb(Const("App", _), func), arg) -> is_quasi_construction func && is_quasi_construction arg
-      | Comb(Comb(Const("Abs", _), var), body) -> is_quasi_construction var && is_quasi_construction body 
+      | Comb(Const("Quo", ty), e) -> is_construction e
+      | Comb(Comb(Const("App", _), func), arg) -> is_construction func && is_construction arg
+      | Comb(Comb(Const("Abs", _), var), body) -> is_construction var && is_construction body 
       | _ when type_of tm = ep_ty -> true 
       | _ -> failwith "Not a quasi construction"
```
Comment (JacquesCarette) [inline fusion.ml]: Shouldn't the error message be consistent?

**Review by JacquesCarette — CHANGES_REQUESTED** [review]

**Review by JacquesCarette — APPROVED** [review]

**Pull request merged**